### ISSUE TITLE
structcheck unused false positive

### ISSUE THREAD

**Issue by SoMuchForSubtlety** [golangci/golangci-lint#1539]
- [x] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [x] Yes, I've searched similar issues on GitHub and didn't find any.
- [x] Yes, I've included all information below (version, config, etc).

Structcheck reports fields as unused due to build constraints. You can see an example here https://github.com/SoMuchForSubtlety/f1viewer/commit/7956d08642e2314e8fd0ed8ade37bdcd5c4a0fc9

`registry` and `registry32` are used if `GOOS==windows`

<details><summary>Version of golangci-lint</summary>

```console
❯ golangci-lint --version
golangci-lint has version v1.33.0 built from (unknown, mod sum: "h1:/o4OtOR3Idim4FHKBJXcy+6ZjNDm82gwK/v6+gWyH9U=") on (unknown)
```

</details>

<details><summary>Config file</summary>

```console
❯ cat .golangci.yml
cat: .golangci.yml: No such file or directory
```

</details>

<details><summary>Go environment</summary>

```console
❯ go version && go env
go version go1.15.5 linux/amd64
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/jakob/.cache/go-build"
GOENV="/home/jakob/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/jakob/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/jakob/go"
GOPRIVATE=""
GOPROXY="direct"
GOROOT="/usr/lib/golang"
GOSUMDB="off"
GOTMPDIR=""
GOTOOLDIR="/usr/lib/golang/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/jakob/projects/f1viewer/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build586356209=/tmp/go-build -gno-record-gcc-switches"
```

</details>

<details><summary>Verbose output of running</summary>

```console
❯ golangci-lint cache clean
❯ golangci-lint run -v
INFO [config_reader] Config search paths: [./ /home/jakob/projects/f1viewer /home/jakob/projects /home/jakob /home /] 
INFO [lintersdb] Active 10 linters: [deadcode errcheck gosimple govet ineffassign staticcheck structcheck typecheck unused varcheck] 
INFO [loader] Go packages loading at mode 575 (types_sizes|compiled_files|deps|exports_file|files|imports|name) took 734.901559ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 10.92354ms 
INFO [linters context/goanalysis] analyzers took 11.035849342s with top 10 stages: buildir: 8.825990025s, inspect: 552.559705ms, fact_purity: 432.376263ms, printf: 377.887282ms, ctrlflow: 377.186351ms, fact_deprecated: 300.815377ms, ineffassign: 16.047312ms, S1038: 13.451378ms, SA1013: 6.132631ms, S1039: 5.965501ms 
INFO [linters context/goanalysis] analyzers took 155.751488ms with top 10 stages: buildir: 108.150613ms, U1000: 47.600875ms 
INFO [runner] Issues before processing: 6, after processing: 2 
INFO [runner] Processors filtering stat (out/in): cgo: 6/6, autogenerated_exclude: 6/6, identifier_marker: 6/6, severity-rules: 2/2, sort_results: 2/2, skip_files: 6/6, exclude-rules: 2/6, uniq_by_line: 2/2, diff: 2/2, path_shortener: 2/2, path_prefixer: 2/2, filename_unadjuster: 6/6, path_prettifier: 6/6, skip_dirs: 6/6, nolint: 2/2, max_from_linter: 2/2, exclude: 6/6, max_per_file_from_linter: 2/2, max_same_issues: 2/2, source_code: 2/2 
INFO [runner] processing took 2.847922ms with stages: nolint: 1.321608ms, identifier_marker: 1.037353ms, autogenerated_exclude: 132.768µs, source_code: 95.752µs, path_prettifier: 95.404µs, exclude-rules: 63.067µs, max_same_issues: 23.606µs, skip_dirs: 16.832µs, uniq_by_line: 8.521µs, cgo: 8.521µs, path_shortener: 6.705µs, max_from_linter: 6.705µs, max_per_file_from_linter: 5.167µs, filename_unadjuster: 4.82µs, diff: 4.329µs, exclude: 3.772µs, skip_files: 3.423µs, sort_results: 3.353µs, severity-rules: 3.352µs, path_prefixer: 2.864µs 
INFO [runner] linters took 4.405924832s with stages: goanalysis_metalinter: 4.146896529s, unused: 255.659086ms 
cmd.go:20:2: `registry` is unused (structcheck)
	registry   string
	^
cmd.go:21:2: `registry32` is unused (structcheck)
	registry32 string
	^
INFO File cache stats: 1 entries of total size 5.1KiB 
INFO Memory: 53 samples, avg is 249.8MB, max is 476.4MB 
INFO Execution took 5.16391085s         
```

</details>

**Comment by iwankgb** [comment]
Have you tried setting GOOS environment variable when running golangci-lint? You seem to run it on Linux so these two structs are not used.

**Comment by SoMuchForSubtlety** [comment]
Yes, it does not report the issue when GOOS is windows. I guess I could run the github action is windows, but if I have fields that are only used on a different system I will have the same issue again.

**Comment by iwankgb** [comment]
I ma afraid it is expected behaviour. A fix could be declaring structs an functions that are used on certain OSes or archs in files that are built on respective OSes and archs. 
See: https://golang.org/cmd/go/#hdr-Build_constraints
